File: 알고리즘 스터디 문제들/Part 6 - DFS와 BFS/문제/연결 요소의 개수.cpp
#include <iostream>
#include <array>
#include <vector>
using namespace std;

array<vector<int>, 1001> graph;
array<bool, 1001> visited = {};

void dfs(int cur)
{
    visited[cur] = true;
    for (size_t i = 0; i < graph[cur].size(); i++)
    {
        if (!visited[graph[cur][i]])
            dfs(graph[cur][i]);
    }
}

int main()
{
    int N, M;
    cin >> N >> M;

    int node1, node2;
    for (size_t i = 0; i < M; i++)
    {
        cin >> node1 >> node2;
        graph[node1].push_back(node2);
        graph[node2].push_back(node1);
    }

    int cnt = 0;
    for (size_t i = 1; i <= N; i++)
    {
        if (!visited[i])
        {
            cnt++;
            dfs(i);
        }
    }
    cout << cnt;
}

File: 알고리즘 스터디 문제들/Part 6 - DFS와 BFS/문제/연결 요소의 개수.py
# 런타임 에러 방지를 위해 제출 시 언어를 PyPy3로 설정해야 한다
from sys import stdin

graph = [[] for i in range(1001)]
visited = [False for i in range(1001)]


def dfs(cur):
    visited[cur] = True
    for node in graph[cur]:
        if (not visited[node]):
            dfs(node)


N, M = map(int, stdin.readline().split())
for i in range(M):
    node1, node2 = map(int, stdin.readline().split())
    graph[node1].append(node2)
    graph[node2].append(node1)
cnt = 0
for i in range(1, N + 1):
    if (not visited[i]):
        cnt += 1
        dfs(i)
print(cnt)
